### PR TITLE
Refactor to enable better usecase testing and use generators to send async events upstream

### PR DESCRIPTION
**Pull request by ryardley**
### What this does: Introduce pattern to make usecases more testable

This extracts each usecase out to it's own file to make them easier to apply unit tests to as moving forward we will need to do this more and more often with complex business rules coming down the pipeline.

This writes a test for the Swap usecase as an example test so that we can follow this pattern moving forward

```ts
// Configure the usecase
export function MyUsecase(config){
  // Return the usecase function
  return (args) => {
    // ...
  }
}
```

### Why do we need this? 
We are not going to be able to cover all our error handling in our end to end tests so we need to start to rely on unit tests to manage testing our usecase testing. Currently to test our usecases we need to mock everything within the feature. With this pattern we mock using only the dependencies we require. This means we can start to move closer to being more reliant on unit testing over end to end testing which will allow us to more quickly iterate on functionality without manual testing.

### What this also does: Push the approve step into to the peg usecase

When Pegging the approve step was external to the usecase which means that the caller of peg needs to remember the business rules around approving funds before pegging assets. This breaks encapsulation as this is something that is always required and is a part of the underlying business logic of the app. It is unreasonable to expect a caller to call this.

Part of the problem was that we would need to react to approval state in the view which is why we had set it up this way. However if we use a generator we can iterate over events coming back from long multistage processes in a rather elegant way:

```ts
// before view knows about business logic 
async function handlePegRequested() {
  const asset = Asset.get(symbol.value);
  const assetAmount = AssetAmount(asset, toBaseUnits(amount.value, asset));

  if (asset.symbol !== "eth") {
    // if not eth you need to approve spend before peg
    transactionState.value = "approving";
    try {
      await usecases.peg.approve(store.wallet.eth.address, assetAmount);
    } catch (err) {
      return (transactionState.value = "rejected");
    }
  }

  transactionState.value = "signing";

  const tx = await usecases.peg.peg(assetAmount);

  transactionHash.value = tx.hash;
  transactionState.value = toConfirmState(tx.state); // TODO: align states
  transactionStateMsg.value = tx.memo ?? "";
}
```

```ts
// now because of async generator view just reacts to async event stream from peg usecase
async function handlePegRequested() {
  const asset = Asset.get(symbol.value);
  const assetAmount = AssetAmount(asset, toBaseUnits(amount.value, asset));

  for await (const event of usecases.peg.peg(assetAmount)) {
    switch (event.type) {
      case "approve_started":
        transactionState.value = "approving";
        break;
      case "approve_error":
        transactionState.value = "rejected";
        break;
      case "signing":
        transactionState.value = "signing";
        break;
      case "sent":
      case "tx_error": {
        const tx = (event as PegSentEvent | PegTxError).tx;
        transactionHash.value = tx.hash;
        transactionState.value = toConfirmState(tx.state); // TODO: align states
        transactionStateMsg.value = tx.memo ?? "";
      }
    }
  }
}
```

### Why we need this?

We need an example of how to read async events from underlying processes. We need to not be afraid of using usecases to maintain complex business logic as that is exactly what they have been setup to accommodate and have been designed to be well unit tested. Secondly this prevents us utilizing an event bus for non global usecase specific events. This means we don't accidentally get global effects from things that we don't expect.

### CODE REVIEW

File: ui/core/src/usecases/clp/addLIquidity.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import { ErrorCode, getErrorMessage, IAssetAmount } from "../../entities";
```
Comment (ajoslin) [inline ui/core/src/usecases/clp/addLIquidity.ts:1]: typo: you put `addLIquidity`

File: ui/core/src/usecases/clp/index.ts [hunk 2]
```diff
@@ -16,193 +12,37 @@ export default ({
...
-      asymmetry: string,
-    ) {
-      const tx = await services.clp.removeLiquidity({
-        fromAddress: state.address,
-        asset,
-        asymmetry,
-        wBasisPoints,
+      effect(() => {
+        // When sif address changes syncPools
+        store.wallet.sif.address;
```
Comment (ajoslin) [inline ui/core/src/usecases/clp/index.ts:39]: This is just a 'read' to signal to the Vue reactive store that we're listening to this field within this effect? A little dicey but I get it.

File: ui/core/src/usecases/peg/peg.test.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    approveBridgeBankSpend: jest.fn().mockReturnValue(Promise.resolve()),
+  };
+  let bus = { dispatch: jest.fn() };
+  let wallet: any = {
+    eth: { address: "eth123", chainId: "0x539" },
+    sif: { address: "sif123" },
+  };
+  let tx: any = {};
+  let subscribeToTx = jest.fn();
+  let SubScribeToTx = () => subscribeToTx;
```
Comment (ajoslin) [inline ui/core/src/usecases/peg/peg.test.ts:20]: SubScribe doesn't make sense, since Subscribe is 1 word

File: ui/core/src/usecases/peg/unpeg.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+      services.bus.dispatch({
+        type: "PegTransactionErrorEvent",
+        payload: {
+          txStatus,
+          message: txStatus.memo || "There was an error while unpegging",
+        },
+      });
+    }
+
+    console.log(
```
Comment (ajoslin) [inline ui/core/src/usecases/peg/unpeg.ts:42]: Console.log still here, intentional?

**Review by ajoslin — COMMENTED** [review]
The patterns make sense to me, seems straightforward to write tests for all of it too

File: ui/core/src/usecases/clp/index.ts (hunk 2 above)
Comment (ryardley) [reply ui/core/src/usecases/clp/index.ts:39]: Yeah I hate this. This is Vue reactivity for you from our current implementation. We should consider the store somewhat deprecated and try to fix this up. Not as part of this refactor though. In my head to fix this we aught to probably isolate the store to be it's own service with event handlers.

File: ui/core/src/usecases/peg/unpeg.ts (hunk 4 above)
Comment (ryardley) [reply ui/core/src/usecases/peg/unpeg.ts:42]: Yeah others like to keep this in.

**Comment by ryardley** [comment]
Lets avoid merging this until after deployment

**Review by thomasdavis — APPROVED** [review]

**Pull request merged**